Filter non-aggregatable fields from vis editor UI

Bargs opened the pull request:
Fixes https://github.com/elastic/kibana/issues/6769 by removing non-aggregatable fields from the vis editor field selector. We get the aggregatable status thanks to https://github.com/elastic/elasticsearch/pull/17980

See commit messages for detailed notes.

Obligatory screenshots:

Notice `agent` is not aggregatable

![screen shot 2016-09-21 at 7 37 38 pm](https://cloud.githubusercontent.com/assets/6239176/18732667/06acf4da-8033-11e6-9177-1298708cd47b.png)

It's not available for aggregation! Only the `raw` version is, as expected

![screen shot 2016-09-21 at 7 39 38 pm](https://cloud.githubusercontent.com/assets/6239176/18732682/398a9dbc-8033-11e6-81a1-71415bc12844.png)

spalger commented on src/core_plugins/kibana/server/routes/api/ingest/register_fields.js:
```diff
@@ -0,0 +1,31 @@
+      const callWithRequest = server.plugins.elasticsearch.callWithRequest;
+      const indices = req.params.indices || '';
+
+      callWithRequest(req, 'fieldStats', {
```
This should probably be returning the promise, in case an error propagates up

spalger commented on src/core_plugins/kibana/server/routes/api/ingest/register_fields.js:
```diff
@@ -0,0 +1,31 @@
+
+export function registerFields(server) {
+  server.route({
+    path: '/api/kibana/{indices}/fields',
```
This feels like a slightly misleading route for what it does.

Bargs replied:
Yeah, I gave it this name because I plan to expand it to provide all the field info Kibana needs once I [remove the mapping cache](https://github.com/elastic/kibana/pull/6648). That said, I agree it could be misleading atm so I'm open to giving it a different name in the interim. Any good ideas? `field_info` maybe? `field_stats` came to mind but seems like a bad idea since I'm filtering out most of the field_stats info.

spalger replied:
`field_usability_stats`? Naming is hard

spalger replied:
`field_es_compatibility`?

Bargs replied:
`field_capabilities`?

Bargs commented:
@spalger I updated the endpoint name and added a mechanism to automatically refresh the field list if `aggregatable` and `searchable` are missing. That was a great catch on the issue with upgrading. Should be ready for another look.

spalger commented on src/ui/public/index_patterns/_index_pattern.js:
```diff
@@ -82,23 +82,33 @@ export default function IndexPatternFactory(Private, Notifier, config, kbnIndex,
-    if (!indexPattern.fields) {
-      return indexPattern.refreshFields();
+
+    let promise = Promise.resolve();
```
Lets move this to the top of the function so that the return on 104 can also return a promise.

thomasneirynck commented:
LGTM, small suggestion about adding some doc for end-user.

Bargs commented:
@spalger @thomasneirynck 

* Tests are passing now
* Added helpful tool tips on the index pattern page
* Updated that function to always return a promise

Let me know what you think

thomasneirynck commented:
LGTMagain. Help text is clear and to the point too!

spalger commented:
👍

LeeDr commented on src/core_plugins/kibana/public/visualize/editor/agg_params.js:
```diff
@@ -138,7 +138,7 @@ uiModules
 
       function getIndexedFields(param) {
-        let fields = $scope.agg.vis.indexPattern.fields.raw;
+        let fields = _.filter($scope.agg.vis.indexPattern.fields.raw, 'aggregatable');
```
I'm guessing that this change may have caused scripted fields to no longer show in the list for metric aggregations?
https://github.com/elastic/kibana/issues/8585
